src/Misc/coroutine.ts

```diff
@@ -0,0 +1,133 @@
+// Coroutines can call other coroutines via:
+// 1. yield* someOtherCoroutine(); // If the called coroutine does not return a value
+// 2. const result = yield* someOtherCoroutine(); // If the called coroutine returns a value
+
+// Coroutines are run with the runCoroutine function, which takes a Coroutine<T>, a CoroutineScheduler<T>, and a success and error callback.
+// A scheduler is responsible for scheduling the next step of a coroutine, either synchronously or asynchronously.
+
+/**
+ * A Coroutine<T> is the intersection of:
+ * 1. An Iterator that yields undefined, returns a T, and is not passed values with calls to next.
+ * 2. An IterableIterator of undefined (since it only yields undefined).
+ */
+export type Coroutine<T> = Iterator<undefined, T, unknown> & IterableIterator<undefined>;
+
+// A CoroutineStep<T> represents a single step of a coroutine, and is an IteratorResult as returned from Coroutine<T>.next().
+type CoroutineStep<T> = IteratorResult<undefined, T>;
+
+// A CoroutineScheduler<T> is responsible for scheduling the call to Coroutine<T>.next and invokes the success or error callback after next is called.
+type CoroutineScheduler<T> = (coroutine: Coroutine<T>, onSuccess: (stepResult: CoroutineStep<T>) => void, onError: (stepError: any) => void) => void;
+
+// The inline scheduler simply steps the coroutine synchronously. This is useful for running a coroutine synchronously, and also as a helper function for other schedulers.
+function inlineScheduler<T>(coroutine: Coroutine<T>, onSuccess: (stepResult: CoroutineStep<T>) => void, onError: (stepError: any) => void) {
+    try {
+        onSuccess(coroutine.next());
+    } catch (error) {
+        onError(error);
+    }
+}
+
+// The yielding scheduler steps the coroutine synchronously until the specified time interval has elapsed, then yields control so other operations can be performed.
+// A single instance of a yielding scheduler could be shared across multiple coroutines to yield when their collective work exceeds the threshold.
+function createYieldingScheduler<T>(yieldAfterMS = 25): CoroutineScheduler<T> {
+    let start: number | undefined;
+    return (coroutine: Coroutine<T>, onSuccess: (stepResult: CoroutineStep<T>) => void, onError: (stepError: any) => void) => {
+        if (start === undefined) {
+            // If start is undefined, this is the first step of a coroutine.
+            start = performance.now();
+        } else {
+            // Otherwise check if the time interval has elapsed.
+            const end = performance.now();
+            if (end - start > yieldAfterMS) {
+                // If so, record a new start time, and schedule the coroutine step to happen later, effectively yielding control of the execution context.
+                start = end;
+                setTimeout(() => {
```

Comment:
queueMicrotask might be better here if available :-)

Reply:
It doesn't work, I think because we need to let control return to the browser to execute `requestAnimationFrame` so the next frame will actually render. If I yield with `queueMicrotask`, no frames are rendered while the coroutine is running.

Reply:
Ohhh if it is to run on next frame, should we use the next render loop instead of a 0 timeout ?

Reply:
We can't rely on the render loop. There might not be a render loop running. If there isn't, this still needs to work.